AccessControl: Introduce new permissions to restrict access for reloading provisioning configuration

vtorosyan opened the pull request:
**What this PR does / why we need it**:

We have introduced new set of fine-grained permissions which allow users to restrict access for reloading provisioning configuration. This PR documents the new permissions and updates certain portion of docs to reflect a rename of one of the permissions related to the Grafana Enterprise.

**Which issue(s) this PR fixes**:

<!--

- Automatically closes linked issue when the Pull Request is merged.

Usage: "Fixes #<issue number>", or "Fixes (paste link of issue)"

-->

https://github.com/grafana/grafana-enterprise/issues/1648

vtorosyan committed:
Access control: Document new permissions restricting access for reloading provisioning configuration

gamab reviewed (approved):
Looks good to me, thanks for taking care of this :100:

osg-grafana commented on docs/sources/enterprise/access-control/permissions.md:
```diff
@@ -23,50 +23,50 @@ scope
+| `reports:send`             | `reports:*`                                                                             | Send a report email.                                                                                                                                     |
+| `reports.settings:write`   | n/a                                                                                     | Update report settings.                                                                                                                                  |
+| `reports.settings:read`    | n/a                                                                                     | Read report settings.                                                                                                                                    |
+| `provisioning:reload`      | `provisioners:*`                                                                        | Reload provisioning files. See [Scope Definitions]({{< relref "./permissions.md#scope-definitions" >}}) to find the exact scope for specific provisioner |
```
```suggestion
| `provisioning:reload`      | `provisioners:*`                                                                        | Reload provisioning files. To find the exact scope for specific provisioner, see [Scope definitions]({{< relref "./permissions.md#scope-definitions" >}}). |
```

osg-grafana commented on docs/sources/http_api/admin.md:
```diff
@@ -612,9 +612,13 @@ Only works with Basic Authentication (username and password). See [introduction]
-| Action              | Scope                  | Provision entity |
-| ------------------- | ---------------------- | ---------------- |
-| provisioning:reload | services:accesscontrol | accesscontrol    |
+| Action              | Scope                      | Provision entity |
```
```suggestion
| Actions              | Scopes                      | Provision entities |
```

osg-grafana commented:
Default to plural headings for consistency. :)

vtorosyan committed:
Apply suggestions from code review

Co-authored-by: Ursula Kallio <73951760+osg-grafana@users.noreply.github.com>

vtorosyan commented:
@osg-grafana would appreciate second look at this. I have enabled auto-merge.

gamab reviewed (changes requested):
Hey, went over this again, I think we are missing the fixed role description and grant :slightly_smiling_face:

vtorosyan committed:
Add fixed role description for provisioning admin

vtorosyan committed:
Add a description

gamab reviewed (approved):
Thanks, sorry for not seeing this earlier !

vtorosyan commented:
@gamab thanks, good catch!

achatterjee-grafana commented on docs/sources/enterprise/access-control/fine-grained-access-control-references.md:
```diff
@@ -15,6 +15,7 @@ The reference information that follows complements conceptual information about
 | ------------------------------ | ---------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- | --------------------------------------------------------------------------------------------- ...
 | `fixed:permissions:admin:read` | `roles:read`<br>`roles:list`<br>`roles.builtin:list`                                                                                                                                                                                                                         | Allows to list and get available roles and built-in role assignments.                         ...
 | `fixed:permissions:admin:edit` | All permissions from `fixed:permissions:admin:read` and <br>`roles:write`<br>`roles:delete`<br>`roles.builtin:add`<br>`roles.builtin:remove`                                                                                                                                 | Allows every read action and in addition allows to create, change and delete custom roles and ...
+| `fixed:provisioning:admin`     | `provisioning:reload`                                                                                                                                                                                                                                                        | Allows to reload provisioning configurations.                                                 ...
```
```suggestion
| `fixed:provisioning:admin`     | `provisioning:reload`                                                                                                                                                                                                                                                        | Allows reload of provisioning configurations.                                                                                             |
```

achatterjee-grafana commented on docs/sources/http_api/admin.md:
```diff
@@ -612,9 +612,13 @@ Only works with Basic Authentication (username and password). See [introduction]
-| Action              | Scope                  | Provision entity |
-| ------------------- | ---------------------- | ---------------- |
-| provisioning:reload | services:accesscontrol | accesscontrol    |
+| Actions              | Scopes                      | Provision entities |
```
Should be singular - Action, Scope, and Provision entity.

osg-grafana replied:
Thank you for the pushback, @vtorosyan. I brought this to the Docs Squad last week to come to a collective decision. The result of which is that we can consistently advise folks on using singular as opposed to plural, by default.

achatterjee-grafana reviewed (commented):
Added doc review.

vtorosyan committed:
Update docs/sources/enterprise/access-control/fine-grained-access-control-references.md

Co-authored-by: achatterjee-grafana <70489351+achatterjee-grafana@users.noreply.github.com>

osg-grafana commented:
> Default to plural headings for consistency. :)

Nope. I brought this to the Docs Squad last week, and we have collectively decided to use singular column names.

osg-grafana commented on docs/sources/enterprise/access-control/permissions.md:
```diff
@@ -23,50 +23,50 @@ scope
+| `org.users.add`            | `users:*`                                                                               | Add a user to an organization.                                                                                                                           |
+| `org.users.remove`         | `users:*`                                                                               | Remove a user from an organization.                                                                                                                      |
+| `org.users.role:update`    | `users:*`                                                                               | Update the organization role (`Viewer`, `Editor`, `Admin`) for an organization.                                                                          |
+| `ldap.user:read`           | n/a                                                                                     | Get a user via LDAP.                                                                                                                                     |
```
I am wondering about this verb and see that the code uses "read" and the plain text uses "get". What is happening here? Asked another way, what is the point of getting/reading a user from LDAP?

vtorosyan replied:
There is currently [another conversation](https://raintank-corp.slack.com/archives/C01LU2Y5527/p1631179555006300) about consistency and naming the roles/permissions . Since this is not something changed in this PR, perhaps we can move that discussion there.

osg-grafana commented on docs/sources/enterprise/access-control/permissions.md:
```diff
@@ -23,50 +23,50 @@ scope
+| `org.users.role:update`    | `users:*`                                                                               | Update the organization role (`Viewer`, `Editor`, `Admin`) for an organization.                                                                          |
+| `ldap.user:read`           | n/a                                                                                     | Get a user via LDAP.                                                                                                                                     |
+| `ldap.user:sync`           | n/a                                                                                     | Sync a user via LDAP.                                                                                                                                    |
+| `ldap.status:read`         | n/a                                                                                     | Verify the LDAP servers’ availability.                                                                                                                   |
```
```suggestion
| `ldap.status:read`         | n/a                                                                                     | Verify the availability of the LDAP server or servers.                                                                                                                   |
```
I assume there can be one or more servers. Vet for accuracy.

osg-grafana commented on docs/sources/enterprise/access-control/permissions.md:
```diff
@@ -23,50 +23,50 @@ scope
+| `ldap.user:sync`           | n/a                                                                                     | Sync a user via LDAP.                                                                                                                                    |
+| `ldap.status:read`         | n/a                                                                                     | Verify the LDAP servers’ availability.                                                                                                                   |
+| `ldap.config:reload`       | n/a                                                                                     | Reload the LDAP configuration.                                                                                                                           |
+| `status:accesscontrol`     | `services:accesscontrol`                                                                | Get access-control enabled status.                                                                                                                       |
```
What is happening here? This sounds a bit cryptic. What is the user’s goal here?

gamab replied:
The goal is to know whether `fine grained access control` is in use. If the status says `disabled`, then, we are not enforcing fined-grained access control, but rather the legacy access control based on the organization role solely.

osg-grafana commented on docs/sources/enterprise/access-control/permissions.md:
```diff
@@ -23,50 +23,50 @@ scope
+| `ldap.status:read`         | n/a                                                                                     | Verify the LDAP servers’ availability.                                                                                                                   |
+| `ldap.config:reload`       | n/a                                                                                     | Reload the LDAP configuration.                                                                                                                           |
+| `status:accesscontrol`     | `services:accesscontrol`                                                                | Get access-control enabled status.                                                                                                                       |
+| `settings:read`            | `settings:*`<br>`settings:auth.saml:*`<br>`settings:auth.saml:enabled` (property level) | Read settings                                                                                                                                            |
```
Please fill in the blank: "Read the _____ ?______ settings."

vtorosyan replied:
Not sure I understand? This line has changed because of automatic formatting

mjseaman replied:
```suggestion
| `settings:read`            | `settings:*`<br>`settings:auth.saml:*`<br>`settings:auth.saml:enabled` (property level) | Read the [Grafana configuration settings]({{< relref "../../administration/configuration/_index.md" >}})                                                                                                                                       |
```

osg-grafana commented on docs/sources/enterprise/access-control/permissions.md:
```diff
@@ -23,50 +23,50 @@ scope
+| `ldap.config:reload`       | n/a                                                                                     | Reload the LDAP configuration.                                                                                                                           |
+| `status:accesscontrol`     | `services:accesscontrol`                                                                | Get access-control enabled status.                                                                                                                       |
+| `settings:read`            | `settings:*`<br>`settings:auth.saml:*`<br>`settings:auth.saml:enabled` (property level) | Read settings                                                                                                                                            |
+| `settings:write`           | `settings:*`<br>`settings:auth.saml:*`<br>`settings:auth.saml:enabled` (property level) | Update settings                                                                                                                                          |
```
Please fill in the blank: "Update the _____ ?______ settings."

mjseaman replied:
```suggestion
| `settings:write`           | `settings:*`<br>`settings:auth.saml:*`<br>`settings:auth.saml:enabled` (property level) | Update any Grafana configuration settings that can be [updated at runtime]({{< relref "../../enterprise/settings-updates/_index.md" >}}).                                                                                                                                           |
```

osg-grafana commented on docs/sources/enterprise/access-control/permissions.md:
```diff
@@ -77,7 +77,8 @@ The following list contains fine-grained access control scopes.
 | `permissions:delegate`   | The scope is only applicable for roles associated with the Access Control itself and indicates that you can delegate your permissions only, or a subset of it, by creating a new role or making an assignment.                                                 |
 | `reports:*`              | Restrict an action to a set of reports. For example, `reports:*` matches any report and `reports:1` matches the report with id `1`.                                                                                                                            |
-| `services:accesscontrol` | Restrict an action to target only the fine-grained access control service. For example, you can use this in conjunction with the `provisioning:reload` or the `status:accesscontrol` actions.                                                                  |
+| `services:accesscontrol` | Restrict an action to target only the fine-grained access control service. For example, you can use this in conjunction with the `status:accesscontrol` actions.                                                                                               |
```
```suggestion
| `services:accesscontrol` | Restrict an action to target only the fine-grained access control service. You can use this in conjunction with the `status:accesscontrol` actions.                                                                                               |
```

osg-grafana commented on docs/sources/enterprise/access-control/permissions.md:
```diff
@@ -77,7 +77,8 @@ The following list contains fine-grained access control scopes.
 | `global:users:*`         | Restrict an action to a set of global users.                                                                                                                                                                                                                   |
 | `users:*`                | Restrict an action to a set of users from an organization.                                                                                                                                                                                                     |
 | `settings:*`             | Restrict an action to a subset of settings. For example, `settings:*` matches all settings, `settings:auth.saml:*` matches all SAML settings, and `settings:auth.saml:enabled` matches the enable property on the SAML settings.                               |
+| `provisioners:*`         | Restrict an action to a set of provisioners. For example, `provisioners:*` matches any provisioner and `provisioners:accesscontrol` matches the Fine-grained access control [provisioner]({{< relref "./provisioning.md" >}}).                                 |
```
```suggestion
| `provisioners:*`         | Restrict an action to a set of provisioners. For example, `provisioners:*` matches any provisioner, and `provisioners:accesscontrol` matches the fine-grained access control [provisioner]({{< relref "./provisioning.md" >}}).                                 |
```

vtorosyan commented on docs/sources/enterprise/access-control/fine-grained-access-control-references.md:
```diff
@@ -30,8 +31,8 @@ The reference information that follows complements conceptual information about
-| Editor         | `fixed:datasource:editor:read`                                                                                                                                                                                                                                                                                                                                                               ...
+| Built-in roles | Associated roles                                                                                                                                                                                                                                                                                                                                                                             ...
+| -------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- ...
+| Grafana Admin  | `fixed:permissions:admin:edit`<br>`fixed:permissions:admin:read`<br>`fixed:provisioning:admin`<br>`fixed:reporting:admin:edit`<br>`fixed:reporting:admin:read`<br>`fixed:users:admin:edit`<br>`fixed:users:admin:read`<br>`fixed:users:org:edit`<br>`fixed:users:org:read`<br>`fixed:ldap:admin:edit`<br>`fixed:ldap:admin:read`<br>`fixed:server:admin:read`<br>`fixed:settings:admin:read` ...
```
Sorry that it's hard to review, I've only added one more fixed permission to the list here and it shows as entire thing has changed.

Anyway, the proposal makes sense to me. 

The only thing is that `Grafana Admin` is the name of the role itself in [Grafana](https://grafana.com/docs/grafana/latest/permissions/organization_roles/#organization-admin-role), so Grafana administrator might not be something users are used to. That said, if this is how we want to go further, happy to make these changes.

vtorosyan commented on docs/sources/enterprise/access-control/permissions.md:
```diff
@@ -23,50 +23,50 @@ scope
-| `datasources:explore`      | n/a                                                                                     | Enable explore                                                                  |
+| Actions                    | Applicable scopes                                                                       | Descriptions                                                                                                                                             |
+| -------------------------- | --------------------------------------------------------------------------------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------- |
+| `roles:list`               | `roles:*`                                                                               | List available roles without permissions.                                                                                                                |
```
I have not changed anything here, but I believe that this role gives access to list the roles and the response does not contain permissions.

vtorosyan committed:
Apply suggestions from code review

Co-authored-by: Ursula Kallio <73951760+osg-grafana@users.noreply.github.com>
Co-authored-by: achatterjee-grafana <70489351+achatterjee-grafana@users.noreply.github.com>

vtorosyan commented:
Thanks @osg-grafana for the review. Because of the automatic formatting, the PR seems much bigger than it is, as I changed only few things here (mainly adding new permissions/roles). I appreciate though the review of the rest, I have added bunch of your suggestions in the change and left some open, which we can hopefully address in a different conversation.

mjseaman commented on docs/sources/enterprise/access-control/permissions.md:
```diff
@@ -23,50 +23,50 @@ scope
+| `org.users.read`           | `users:*`                                                                               | Get user profiles within an organization.                                                                                                                |
+| `org.users.add`            | `users:*`                                                                               | Add a user to an organization.                                                                                                                           |
+| `org.users.remove`         | `users:*`                                                                               | Remove a user from an organization.                                                                                                                      |
+| `org.users.role:update`    | `users:*`                                                                               | Update the organization role (`Viewer`, `Editor`, or `Administrator`) of an organization.                                                                          |
```
```suggestion
| `org.users.role:update`    | `users:*`                                                                               | Update the organization role (`Viewer`, `Editor`, or `Admin`) of an organization.                                                                          |
```

mjseaman committed:
Update docs/sources/enterprise/access-control/permissions.md

vtorosyan committed:
Apply suggestions from code review

Co-authored-by: Mitch Seaman <mjseaman@users.noreply.github.com>

vtorosyan commented:
Thanks everyone for contributions. I'd appreciate if we can avoid increasing the scope of this PR. The automatic formatting resulted to many lines being shown as updated, however the scope of this PR includes only adding new permissions/roles. 

To be clear, I am not suggesting to leave the status-quo, rather let's give another go over docs and follow-up with more PRs. As we are updating docs in few places, the more we add here the harder it becomes to sync those changes together later .

sakjur commented on docs/sources/enterprise/access-control/fine-grained-access-control-references.md:
```diff
@@ -30,8 +31,8 @@ The reference information that follows complements conceptual information about
-| Editor         | `fixed:datasource:editor:read`                                                                                                                                                                                                                                                                                                                                                               ...
+| Built-in role | Associated role                                                                                                                                                                                                                                                                                                                                                                               ...
+| -------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- ...
+| Grafana administrator  | `fixed:permissions:admin:edit`<br>`fixed:permissions:admin:read`<br>`fixed:provisioning:admin`<br>`fixed:reporting:admin:edit`<br>`fixed:reporting:admin:read`<br>`fixed:users:admin:edit`<br>`fixed:users:admin:read`<br>`fixed:users:org:edit`<br>`fixed:users:org:read`<br>`fixed:ldap:admin:edit`<br>`fixed:ldap:admin:read`<br>`fixed:server:admin:read`<br>`fixed:settings:adm ...
+| Administrator          | `fixed:users:org:edit`<br>`fixed:users:org:read`<br>`fixed:reporting:admin:edit`<br>`fixed:reporting:admin:read`                                                                                                                                                                                                                                                                     ...
```
They're assigned to the strings `"Admin"` and `"Grafana Admin"` respectively. Do we have anything that makes it clear that that's when you want to change the organizational administrator's permissions, you should provision for `Admin` rather than `Administrator` or should that too be introduced? (not in this PR, regardless. We shouldn't let PRs be blocked by work that's not directly related to the PR)

vtorosyan replied:
The [provisioning docs](https://grafana.com/docs/grafana/latest/enterprise/access-control/provisioning/#built-in-role-assignments) and [API docs](https://grafana.com/docs/grafana/latest/http_api/access_control/#create-a-built-in-role-assignment) state this clearly. 

In other places it is referencing to the current docs where it's not very clear in either way. I'm not against this change (so I accepted the suggestion), however I agree that it won't make things less confusing for users. So if this is a one step forward towards consistent glossary and eliminates future work, I'm fine to update. But @osg-grafana would appreciate your call to unblock this PR, we can take this conversation to private and understand how we can move forward with FGAC docs.

sakjur replied:
That's good enough for me :+1: Thanks for the links

osg-grafana commented on docs/sources/enterprise/access-control/fine-grained-access-control-references.md:
```diff
@@ -30,8 +31,8 @@ The reference information that follows complements conceptual information about
+| Built-in role | Associated role                                                                                                                                                                                                                                                                                                                                                                               ...
+| -------------- | -------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------- ...
+| Grafana administrator  | `fixed:permissions:admin:edit`<br>`fixed:permissions:admin:read`<br>`fixed:provisioning:admin`<br>`fixed:reporting:admin:edit`<br>`fixed:reporting:admin:read`<br>`fixed:users:admin:edit`<br>`fixed:users:admin:read`<br>`fixed:users:org:edit`<br>`fixed:users:org:read`<br>`fixed:ldap:admin:edit`<br>`fixed:ldap:admin:read`<br>`fixed:server:admin:read`<br>`fixed:settings:adm ...
+| Administrator          | `fixed:users:org:edit`<br>`fixed:users:org:read`<br>`fixed:reporting:admin:edit`<br>`fixed:reporting:admin:read`                                                                                                                                                                                                                                                                     ...
```
```suggestion
| Admin          | `fixed:users:org:edit`<br>`fixed:users:org:read`<br>`fixed:reporting:admin:edit`<br>`fixed:reporting:admin:read`                                                                                                                                                                                                                                                                                                            | Allow access to the same resources and permissions that the [Organization Admin]({{< relref "../../permissions/organization_roles.md" >}}) has by default.                               |
```

osg-grafana reviewed (approved):
I think I successfully revert a couple of my original suggestions.

vtorosyan committed:
Apply suggestions from code review

Co-authored-by: Ursula Kallio <73951760+osg-grafana@users.noreply.github.com>
